Watch-only CN of 4-nodes privnet can't participate in TPKE-based block construction

#### System information

Commit hash: 438fa77fe (current bane-main)

#### Actual behaviour

Watch-only CN node from *4-nodes* privnet can't build new blocks starting from `NeoXAMEV-1` height (34 for privnet) due to invalid NextConsensus calculation:
```
INFO [01-14|14:54:20.474] New block in the chain                   "dbft index"=33 "chain index"=33 hash=0xcc455cb4e8df61039c92e7c3d895e1f7f6c765ac065bc4f723bbe88c68ce5d76 "parent hash"=0xe22025b948020260548691ba09210fcc26b13a464e275f4ef0ea689a2ba30b74 primary=1 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0xce2c47e402f65838a2251f779218a3e69debb1236a1099d6fb7e65c04a7065f7
INFO [01-14|14:54:20.474] Fetching latest sealing proposal         "desired number"=34
INFO [01-14|14:54:20.474] Commit new sealing work                  number=34 sealhash=d8d6d3..1da30d "parent hash"=cc455c..ce5d76 etherbase=0x1212000000000000000000000000000000000003 txs=0 gas=0         fees=0          elapsed="654.97µs"
INFO [01-14|14:54:20.478] Imported new chain segment               number=33 hash=b7c710..0ad9f3 blocks=1 txs=0 mgas=0.000 elapsed=1.507ms     mgasps=0.000   snapdiffs=14.93KiB triedirty=93.95KiB
INFO [01-14|14:54:21.475] Sealing proposal updated                 number=34 sealhash=d8d6d3..1da30d "parent hash"=cc455c..ce5d76 txs=0
2025-01-14T14:54:21.475+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:112	initializing dbft	{"height": 34, "view": 0, "index": -1, "role": "WatchOnly"}
2025-01-14T14:54:21.475+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "Commit", "from": 2, "height": 33, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:21.475+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:240	ignoring old height	{"height": 33}
2025-01-14T14:54:25.478+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "PrepareRequest", "from": 2, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
INFO [01-14|14:54:25.479] New parent stored                        number=33 "old hash"=0xcc455cb4e8df61039c92e7c3d895e1f7f6c765ac065bc4f723bbe88c68ce5d76 "new hash"=0xb7c710a772c0a115b43fdbf904ea303f89f7d39d40934447dff91d82560ad9f3 sealhash=0xf185e096f91a4de341942a07769061208e54b7b826b7ddff2393836cdcf447b9 "old extra"=0074f4effb0b538baec703346b03b6d9292f53a4cd910ad1641b7125eff746accdca1f11148b22f472c51964013acbc6b271feecb0febd9e7a01202930fef5f250af14df73f983caab7b1f5002189c42e0a08c1f212d59262363e9fe360e575bf19e7f071193e2acc528729e8c92e8dfa63745a71225e5fbc6a2271c7bb76ff0693ac8120380ed67e653ccfcf1cdfec27e013426b41abec9aff99203093410370629c4a47149366241233b8c9f8eaca9ae135e2631c4337554a59c15f92f6416ca0c60429873defff1edef9f0105e02106fd018630f7572ac8c25aa21e4850f20bcfae70613e692b0a555b79ca57813f11413e12db07c5f3c7b44e868a229f2616b8dd02bb2a7b4c19d483aadf1e2b4412855b00 "new extra"=0074f4effb0b538baec703346b03b6d9292f53a4cd910ad1641b7125eff746accdca1f11148b22f472c51964013acbc6b271feecb0febd9e7a01202930fef5f250af14df73f983caab7b1f5002189c42e0a08c1f212d59262363e9fe360e575bf19e7f071193e2acc528729e8c92e8dfa63745a71225e5fbc6a2271c7bb76ff0693ac8120380ed67e653ccfcf1cdfec27e0195455fa0a060cdcd839bc0c4d91c6fdde5f76e488253d0f8172dc26cd454ac406e4cf7a6fcd16c66d4b112f87d357eb7ebe39d2c01d513f5f589f919c1b492a8008630f7572ac8c25aa21e4850f20bcfae70613e692b0a555b79ca57813f11413e12db07c5f3c7b44e868a229f2616b8dd02bb2a7b4c19d483aadf1e2b4412855b00
2025-01-14T14:54:25.479+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:325	received PrepareRequest	{"validator": 2, "tx": 0}
WARN [01-14|14:54:25.479] Invalid NextConsensus in the proposed block expected=0xb88ec11fde916b6883a5054c24abaaf124772e97e523e1d841c233a55e8b88ba actual=0x2183e53979588941263e03b32a89b43c2d130827688d72cc224679569d05dc65
2025-01-14T14:54:25.479+0300	WARN	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:373	proposed block fails verification
github.com/nspcc-dev/dbft.(*DBFT[...]).createAndCheckBlock
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:373
github.com/nspcc-dev/dbft.(*DBFT[...]).onPrepareRequest
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:330
github.com/nspcc-dev/dbft.(*DBFT[...]).OnReceive
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:268
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1946
2025-01-14T14:54:25.479+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "PrepareResponse", "from": 3, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.479+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:464	received PrepareResponse	{"validator": 3}
2025-01-14T14:54:25.480+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "PrepareResponse", "from": 0, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.480+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:464	received PrepareResponse	{"validator": 0}
2025-01-14T14:54:25.480+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "PrepareResponse", "from": 1, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.480+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:464	received PrepareResponse	{"validator": 1}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "Commit", "from": 0, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.481+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:595	received Commit	{"validator": 0}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/check.go:116	not enough to commit	{"count": 1}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "PrepareResponse", "from": 1, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:452	ignoring PrepareResponse	{"dup": true, "sor": true, "viewChanging": false, "moreThanF": false}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "Commit", "from": 3, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.481+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:595	received Commit	{"validator": 3}
2025-01-14T14:54:25.481+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/check.go:116	not enough to commit	{"count": 2}
2025-01-14T14:54:25.482+0300	DEBUG	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:231	received message	{"type": "Commit", "from": 1, "height": 34, "view": 0, "my_height": 34, "my_view": 0}
2025-01-14T14:54:25.482+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:595	received Commit	{"validator": 1}
2025-01-14T14:54:25.482+0300	INFO	dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/check.go:123	approving block	{"height": 34, "hash": "0xc230bcbc9ee2c45c1171561bf83d9c774335f5e33af81119c1434be869d13d57", "tx_count": 0, "merkle": "0x73f616555412f18003bd39f93be4d16035fc8c18aead22cec1ee4254f3acb971", "prev": "0xb7c710a772c0a115b43fdbf904ea303f89f7d39d40934447dff91d82560ad9f3"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd49205]

goroutine 4311 [running]:
github.com/ethereum/go-ethereum/core/state.(*StateDB).Copy(0x0)
	github.com/ethereum/go-ethereum/core/state/statedb.go:691 +0x25
github.com/ethereum/go-ethereum/consensus/dbft.New.func5({0x1fd8fe0?, 0xc000dc5080})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:377 +0x21c
github.com/nspcc-dev/dbft.(*DBFT[...]).checkCommit(0x1ff8ba0)
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/check.go:130 +0x8d7
github.com/nspcc-dev/dbft.(*DBFT[...]).onCommit(0x1ff8ba0, {0x1fdd870, 0xc001ba45a0})
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:601 +0xd6d
github.com/nspcc-dev/dbft.(*DBFT[...]).OnReceive(0x1ff8ba0, {0x1fdd870, 0xc001ba45a0})
	github.com/nspcc-dev/dbft@v0.3.2-0.20241205125914-b2ba0cd4b763/dbft.go:272 +0xfe5
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).eventLoop(0xc00057afc8)
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1946 +0xb1d
created by github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start in goroutine 4388
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1780 +0x478
```
I checked dBFT-related part, and it seems that everything is OK with the NextConsensus-related code. 7-nodes privnet don't have this problem. The only possible source of the problem to me is anti-MEV keystore setup. Something is wrong with threshold public key on WatchOnly CN by the moment of 34-th block construction which leads to an invalid NextConsensus produced by this node.

#### Expected behaviour

No error on attempt to calculate NextConsensus for `NeoXAMEV-1` block. Anti-MEV keystore should be properly setup and ready-to-use by the moment `NeoXAMEV` is enabled, because `NeoXDKG` is enabled starting from height `5`.

#### Steps to reproduce the behaviour

Run 4-nodes privnet, wait for 35 blocks to be accepted, check node0 logs.

#### Desired solution

As a part of the desired solution I'm requesting for not only CN0 wallet regeneration, but also for a unit-test that should regenerate and setup privnet amev keystore of all CNs. The test itself should be similar to `TestTPKE` and the overall logic should be similar to https://github.com/bane-labs/go-ethereum/pull/326#discussion_r1806237317: based on some private key we need to generate and save keystore to respective `privnet/.../nodeX/antimev-keystore` files for all CNs for all privnets. As a developer, I should be able to run this test and regenerate privnet antimev wallets based on static private key.

Otherwise I don't have a clue what is the source of these data and it's harder to check what is the source of such block construction problem.

--- a/antimev/keygroup.go
+++ b/antimev/keygroup.go
@@ -45,6 +45,31 @@ func (tkg *thresholdKeyGroup) newTemplateForReshare(size int) *thresholdKeyGroup
 	}
 }
 
+// copy creates a deep copy of thresholdKeyGroup.
+func (tkg *thresholdKeyGroup) copy() *thresholdKeyGroup {
+	res := new(thresholdKeyGroup)
+	if tkg.localSecret != nil {
+		res.localSecret = tkg.localSecret.Copy()
+	}
+	res.pendingSecrets = make([]*tpke.Secret, len(tkg.pendingSecrets))
+	for i := range tkg.pendingSecrets {
+		res.pendingSecrets[i] = tkg.pendingSecrets[i].Copy()
+	}
+	res.receivedSecrets = make([]*big.Int, len(tkg.receivedSecrets))
+	for i, secret := range tkg.receivedSecrets {
+		if secret != nil {
+			res.receivedSecrets[i] = new(big.Int).Set(secret)
+		}
+	}
+	if tkg.globalPubKey != nil {
+		res.globalPubKey = tkg.globalPubKey.Copy()
+	}
+	if tkg.localPrvKey != nil {
+		res.localPrvKey = tkg.localPrvKey.Copy()
+	}
+	return res
+}
+
 // prepare generates local secrets and returns sharing messages.
 func (tkg *thresholdKeyGroup) prepare(threshold int, messagePubKeys []*ecies.PublicKey) ([][]byte, *tpke.PVSS, error) {
 	// Generate local secret

--- a/antimev/keygroup.go
+++ b/antimev/keygroup.go
@@ -14,6 +14,7 @@ var (
 	ErrDKGSecret            = errors.New("invalid dkg secret")
 	ErrNoSecretToReshare    = errors.New("no secret to reshare")
 	ErrInvalidRecover       = errors.New("invalid recover")
+	ErrInvalidMessageKey    = errors.New("invalid message key")
 	ErrSecretShareNotEnough = errors.New("secret share not enough")
 )
 
@@ -118,6 +119,9 @@ func (tkg *thresholdKeyGroup) recover(secretIndexs []int, receiverEthPubKeys []*
 	srms := make([][]byte, len(secretIndexs))
 	for i, index := range secretIndexs {
 		arrIndex := index - 1
+		if tkg.receivedSecrets[arrIndex] == nil {
+			return nil, ErrUnrecoverable
+		}
 		srms[i] = encryptShareMessage(receiverEthPubKeys[i], tkg.receivedSecrets[arrIndex].Bytes())
 	}
 
@@ -165,6 +169,9 @@ func generateShareMessages(secret *tpke.Secret, messagePubkeys []*ecies.PublicKe
 	// Generate message
 	messages := make([][]byte, size)
 	for i, key := range messagePubkeys {
+		if key == nil {
+			return nil, nil, ErrInvalidMessageKey
+		}
 		messages[i] = encryptShareMessage(key, ss[i].Bytes())
 	}
 

--- a/antimev/keystore.go
+++ b/antimev/keystore.go
@@ -4,7 +4,6 @@ import (
 	"encoding/hex"
 	"encoding/json"
 	"errors"
-	"fmt"
 	"math/big"
 	"path/filepath"
 
@@ -80,31 +79,36 @@ func (ks *KeyStore) Reset(round int) error {
 	return ks.saveStoreAndReInitialize()
 }
 
-// Copy creates a deep copy of KeyStore.
-func (ks *KeyStore) Copy() (*KeyStore, error) {
-	res := new(KeyStore)
-	res.path = ks.path
-	res.password = ks.password
-
-	// A tiny hack implemented for the current keystore implementation consistency
-	// until a proper persistence functionality is implemented over KeyStore, ref.
-	// https://github.com/bane-labs/go-ethereum/issues/388.
-	err := res.initializeKeystoreFromFile()
-	if err != nil {
-		return nil, fmt.Errorf("failed to initialize keystore from file: %w", err)
-	}
-
-	// Explicitly set path to empty as a temporary replacement of proper persistence
-	// functionality, ref. https://github.com/bane-labs/go-ethereum/issues/388.
-	res.path = ""
-
-	return res, nil
-}
-
 // Load loads hex-encoded anti-MEV keystore from the provided filepath.
 func (ks *KeyStore) Load(password string) error {
 	ks.password = password
 	return ks.initializeKeystoreFromFile()
+}
+
+// Copy creates a deep copy of KeyStore.
+func (ks *KeyStore) Copy() *KeyStore {
+	cp := *ks
+	res := &cp
+	if ks.ethPrvKey != nil {
+		k := *ks.ethPrvKey
+		res.ethPrvKey = &k
+	}
+	if ks.recovering != nil {
+		res.recovering = ks.recovering.copy()
+	}
+	if ks.resharing != nil {
+		res.resharing = ks.resharing.copy()
+	}
+	if ks.reshared != nil {
+		res.reshared = ks.reshared.copy()
+	}
+	if ks.sharing != nil {
+		res.sharing = ks.sharing.copy()
+	}
+	if ks.shared != nil {
+		res.shared = ks.shared.copy()
+	}
+	return res
 }
 
 // Update changes the passphrase of anti-MEV keystore

--- a/antimev/keystore.go
+++ b/antimev/keystore.go
@@ -224,22 +224,19 @@ func (ks *KeyStore) OnRecoverPeriodStart() {
 // array should be ordered by DKG index, e.g. keystore will use 1 as the index
 // to generate a reshare message for messagePubKeys[0].
 func (ks *KeyStore) DKGReshare(messagePubKeys []*ecies.PublicKey) ([][]byte, []byte, error) {
-	if ks.resharing == nil {
+	if ks.shared == nil {
 		return nil, nil, ErrKeyGroupNotExists
 	}
 	if len(messagePubKeys) != ks.size {
 		return nil, nil, ErrInvalidLength
 	}
-	if ks.shared != nil {
-		// Generate secret resharing messages and pvss
-		rMsgs, rPvss, err := ks.shared.reshare(messagePubKeys)
-		if err != nil {
-			return nil, nil, err
-		}
-		// No data changed so no need to persist
-		return rMsgs, rPvss.Encode(), nil
+	// Generate secret resharing messages and pvss
+	rMsgs, rPvss, err := ks.shared.reshare(messagePubKeys)
+	if err != nil {
+		return nil, nil, err
 	}
-	return nil, nil, nil
+	// No data changed so no need to persist
+	return rMsgs, rPvss.Encode(), nil
 }
 
 // DKGShare generates and returns sharing messages and pvss. The input key array
@@ -254,7 +251,7 @@ func (ks *KeyStore) DKGShare(messagePubKeys []*ecies.PublicKey) ([][]byte, []byt
 	}
 	// Generate secret sharing messages and pvss
 	sMsgs, sPvss, err := ks.sharing.prepare(ks.threshold, messagePubKeys)
-	if err != nil || sPvss == nil {
+	if err != nil {
 		return nil, nil, err
 	}
 	return sMsgs, sPvss.Encode(), nil
@@ -264,8 +261,11 @@ func (ks *KeyStore) DKGShare(messagePubKeys []*ecies.PublicKey) ([][]byte, []byt
 // should be a DKG index starts from 1, and the size of two inputs should be the
 // same.
 func (ks *KeyStore) DKGRecover(indexes []int, messagePubKeys []*ecies.PublicKey) ([][]byte, error) {
-	if ks.shared == nil || ks.recovering == nil {
+	if ks.shared == nil {
 		return nil, ErrKeyGroupNotExists
+	}
+	if len(indexes) != len(messagePubKeys) {
+		return nil, ErrLengthMismatch
 	}
 	if len(messagePubKeys) < 1 {
 		return nil, ErrNoNeedToRecover
@@ -283,11 +283,11 @@ func (ks *KeyStore) DKGRecover(indexes []int, messagePubKeys []*ecies.PublicKey)
 // ordered by DKG index, e.g. keystore will use 1 as the index to generate a
 // share message for messagePubKeys[0].
 func (ks *KeyStore) TryRecoverReshare(messagePubKeys []*ecies.PublicKey) ([][]byte, []byte, error) {
-	if ks.recovering == nil || ks.resharing == nil {
+	if ks.recovering == nil {
 		return nil, nil, ErrKeyGroupNotExists
 	}
 	msgs, pvss, err := ks.recovering.reshareRecovered(ks.threshold, messagePubKeys)
-	if err != nil || pvss == nil {
+	if err != nil {
 		return nil, nil, err
 	}
 	return msgs, pvss.Encode(), nil
@@ -392,7 +392,7 @@ func (ks *KeyStore) ReceiveSecretShare(selfIndex int, fromIndex int, ess [][]byt
 // Both selfIndex and fromIndex should be a DKG index starts from 1.
 func (ks *KeyStore) ReceiveSecretReshare(selfIndex int, fromIndex int, ers [][]byte, pvss []byte) error {
 	// Check if out of the period
-	if ks.shared == nil || ks.resharing == nil {
+	if ks.resharing == nil {
 		return ErrKeyGroupNotExists
 	}
 	// Check if the secret share has a valid length
@@ -420,7 +420,7 @@ func (ks *KeyStore) ReceiveSecretReshare(selfIndex int, fromIndex int, ers [][]b
 // Both selfIndex and fromIndex should be a DKG index starts from 1.
 func (ks *KeyStore) ReceiveRecoverShare(selfIndex int, fromIndex int, ers []byte, pvss []byte) error {
 	// Check if out of the period
-	if ks.shared == nil || ks.recovering == nil {
+	if ks.recovering == nil {
 		return ErrKeyGroupNotExists
 	}
 	// Decode PVSS

--- a/antimev/keystore.go
+++ b/antimev/keystore.go
@@ -25,6 +25,9 @@ var (
 	ErrNoNeedToRecover   = errors.New("no need to recover")
 	ErrUnrecoverable     = errors.New("unrecoverable sharing")
 	ErrMessageDecryption = errors.New("message decryption failed")
+
+	ErrKeystoreEncryption = errors.New("could not encrypt keystore")
+	ErrKeystoreDecryption = errors.New("could not decrypt keystore")
 )
 
 // KeyStore is the container of all useful dkg information.
@@ -65,24 +68,46 @@ func (ks *KeyStore) Init(addr common.Address, prvkey *ecies.PrivateKey, groupSiz
 	ks.ethPrvKey = prvkey
 	ks.password = password
 	ks.round = 0
-	return ks.saveStoreAndReInitialize()
+	return nil
+}
+
+// UpdatePassword changes the passphrase
+func (ks *KeyStore) UpdatePassword(password string) {
+	ks.password = password
 }
 
 // Reset cleans all dkg progress data and returns to an initial state
-func (ks *KeyStore) Reset(round int) error {
+func (ks *KeyStore) Reset(round int) {
 	ks.round = round
 	ks.recovering = nil
 	ks.resharing = nil
 	ks.reshared = nil
 	ks.sharing = nil
 	ks.shared = nil
-	return ks.saveStoreAndReInitialize()
 }
 
 // Load loads hex-encoded anti-MEV keystore from the provided filepath.
 func (ks *KeyStore) Load(password string) error {
 	ks.password = password
-	return ks.initializeKeystoreFromFile()
+	encoded, err := readFileAtPath(filepath.Dir(ks.path), filepath.Base(ks.path))
+	if err != nil {
+		return err
+	}
+	keystoreFile := &keystoreRepresentation{}
+	if err := json.Unmarshal(encoded, keystoreFile); err != nil {
+		return err
+	}
+	// Extract the validator signing private key from the keystore
+	decryptor := keystorev4.New()
+	enc, err := decryptor.Decrypt(keystoreFile.Crypto, ks.password)
+	if err != nil {
+		return ErrKeystoreDecryption
+	}
+	store := &keyStoreAux{}
+	if err := json.Unmarshal(enc, store); err != nil {
+		return err
+	}
+	return ks.fromAux(store)
 }
 
 // Copy creates a deep copy of KeyStore.
@@ -111,10 +136,17 @@ func (ks *KeyStore) Copy() *KeyStore {
 	return res
 }
 
-// Update changes the passphrase of anti-MEV keystore
-func (ks *KeyStore) Update(password string) error {
-	ks.password = password
-	return ks.saveStoreAndReInitialize()
+// Persist saves current keystore data to the provided filepath.
+func (ks *KeyStore) Persist() error {
+	keystoreRepr, err := createKeystoreRepresentation(ks.toAux(), ks.password)
+	if err != nil {
+		return err
+	}
+	encodedData, err := json.MarshalIndent(keystoreRepr, "", "\t")
+	if err != nil {
+		return err
+	}
+	return writeFileAtPath(filepath.Dir(ks.path), filepath.Base(ks.path), encodedData)
 }
 
 // Address returns the keystore specified participant address in DKG
@@ -127,15 +159,6 @@ func (ks *KeyStore) Path() (string, error) {
 	return filepath.Abs(ks.path)
 }
 
-// GlobalPublicKey returns global public key that may be used to verify threshold
-// signature against. Do not modify the return value.
-func (ks *KeyStore) GlobalPublicKey() (*tpke.PublicKey, error) {
-	if ks.shared == nil {
-		return nil, ErrNoPubKey
-	}
-	return ks.shared.globalPubKey, nil
-}
-
 // MessagePubKey returns a hex string of message encryption key
 func (ks *KeyStore) MessagePubKey() string {
 	return hex.EncodeToString(crypto.FromECDSAPub(&ks.ethPrvKey.ExportECDSA().PublicKey))
@@ -146,17 +169,17 @@ func (ks *KeyStore) Round() int {
 	return ks.round
 }
 
-// CurrentGlobalPubKey returns global public key that may be used to verify threshold
+// GlobalPublicKey returns global public key that may be used to verify threshold
 // signature against. Do not modify the return value.
-func (ks *KeyStore) CurrentGlobalPubKey() (*tpke.PublicKey, error) {
+func (ks *KeyStore) GlobalPublicKey() (*tpke.PublicKey, error) {
 	if ks.shared == nil {
 		return nil, ErrNoPubKey
 	}
 	return ks.shared.globalPubKey, nil
 }
 
-// LastGlobalPubKey returns last round global public key.
-func (ks *KeyStore) LastGlobalPubKey() (*tpke.PublicKey, error) {
+// LastGlobalPublicKey returns last round global public key.
+func (ks *KeyStore) LastGlobalPublicKey() (*tpke.PublicKey, error) {
 	if ks.reshared == nil {
 		return nil, ErrNoPubKey
 	}
@@ -178,19 +201,9 @@ func (ks *KeyStore) IsRecovering() bool {
 	return ks.recovering != nil
 }
 
-// HasReshared returns if there is a completed resharing
-func (ks *KeyStore) HasReshared() bool {
-	return ks.reshared != nil
-}
-
-// HasShared returns if there is a completed sharing
-func (ks *KeyStore) HasShared() bool {
-	return ks.shared != nil
-}
-
 // OnValidatorList initializes sharing and resharing, should be called
 // when the key group members are determined. It initializes 1 or 2 key groups.
-func (ks *KeyStore) OnSharePeriodStart() error {
+func (ks *KeyStore) OnSharePeriodStart() {
 	// Set up groups for dkg resharing
 	if ks.shared != nil {
 		ks.resharing = ks.shared.newTemplateForReshare(ks.size)
@@ -199,17 +212,12 @@ func (ks *KeyStore) OnSharePeriodStart() error {
 	ks.sharing = newThresholdKeyGroup(ks.size)
 	// Remove recovering just in case
 	ks.recovering = nil
-	return ks.saveStoreAndReInitialize()
 }
 
 // OnRecoverPeriodStart initializes recovering, should be called after resharing
 // gets finished. It initializes 1 key groups.
-func (ks *KeyStore) OnRecoverPeriodStart() error {
-	if ks.shared == nil {
-		return ErrKeyGroupNotExists
-	}
+func (ks *KeyStore) OnRecoverPeriodStart() {
 	ks.recovering = newThresholdKeyGroup(ks.size)
-	return ks.saveStoreAndReInitialize()
 }
 
 // DKGReshare generates and returns resharing messages and pvss. The input key
@@ -249,11 +257,6 @@ func (ks *KeyStore) DKGShare(messagePubKeys []*ecies.PublicKey) ([][]byte, []byt
 	if err != nil || sPvss == nil {
 		return nil, nil, err
 	}
-	// Store only if some secret is generated
-	err = ks.saveStoreAndReInitialize()
-	if err != nil {
-		return nil, nil, err
-	}
 	return sMsgs, sPvss.Encode(), nil
 }
 
@@ -287,20 +290,14 @@ func (ks *KeyStore) TryRecoverReshare(messagePubKeys []*ecies.PublicKey) ([][]by
 	if err != nil || pvss == nil {
 		return nil, nil, err
 	}
-	// Store only if some secret is recovered
-	err = ks.saveStoreAndReInitialize()
-	if err != nil {
-		return nil, nil, err
-	}
 	return msgs, pvss.Encode(), nil
 }
 
 // RevertRound reverts all progress of current round of DKG.
-func (ks *KeyStore) RevertRound() error {
+func (ks *KeyStore) RevertRound() {
 	ks.resharing = nil
 	ks.sharing = nil
 	ks.recovering = nil
-	return ks.saveStoreAndReInitialize()
 }
 
 // OnEpochChange checks and settles the results of sharing and resharing, will
@@ -308,7 +305,8 @@ func (ks *KeyStore) RevertRound() error {
 func (ks *KeyStore) OnEpochChange(selfPvss []byte, aggregatedCmt []byte, isMemberOfNewGroup bool) error {
 	// Revert dkg if contract doesn't give a aggregated commitment
 	if len(aggregatedCmt) == 0 {
-		return ks.RevertRound()
+		ks.RevertRound()
+		return nil
 	}
 	if isMemberOfNewGroup && len(selfPvss) == 0 {
 		return ErrLengthMismatch
@@ -330,7 +328,7 @@ func (ks *KeyStore) OnEpochChange(selfPvss []byte, aggregatedCmt []byte, isMembe
 	ks.sharing = nil
 	ks.recovering = nil
 	ks.round += 1
-	return ks.saveStoreAndReInitialize()
+	return nil
 }
 
 // aggregateShare tries to check if sharing is successful and settle the result
@@ -387,7 +385,7 @@ func (ks *KeyStore) ReceiveSecretShare(selfIndex int, fromIndex int, ess [][]byt
 	if err != nil {
 		return err
 	}
-	return ks.saveStoreAndReInitialize()
+	return nil
 }
 
 // ReceiveSecretReshare tries to verify a resharing message array and store their data.
@@ -415,7 +413,7 @@ func (ks *KeyStore) ReceiveSecretReshare(selfIndex int, fromIndex int, ers [][]b
 	if err != nil {
 		return err
 	}
-	return ks.saveStoreAndReInitialize()
+	return nil
 }
 
 // ReceiveRecoverShare tries to verify a recovering message array and store their data.
@@ -439,7 +437,7 @@ func (ks *KeyStore) ReceiveRecoverShare(selfIndex int, fromIndex int, ers []byte
 	if err != nil {
 		return err
 	}
-	return ks.saveStoreAndReInitialize()
+	return nil
 }
 
 func (ks *KeyStore) decryptShareMessage(msg []byte) (*big.Int, error) {
@@ -581,67 +579,6 @@ func (ks *KeyStore) fromAux(aux *keyStoreAux) error {
 	}
 
 	return nil
-}
-
-var (
-	ErrKeystoreEncryption = errors.New("could not encrypt keystore")
-	ErrKeystoreDecryption = errors.New("could not decrypt keystore")
-)
-
-// initializeKeystoreFromFile loads an existing keystore from file.
-func (ks *KeyStore) initializeKeystoreFromFile() error {
-	encoded, err := readFileAtPath(filepath.Dir(ks.path), filepath.Base(ks.path))
-	if err != nil {
-		return err
-	}
-
-	return ks.initializeKeystoreFromBytes(encoded)
-}
-
-func (ks *KeyStore) initializeKeystoreFromBytes(encoded []byte) error {
-	keystoreFile := &keystoreRepresentation{}
-	if err := json.Unmarshal(encoded, keystoreFile); err != nil {
-		return err
-	}
-	// Extract the validator signing private key from the keystore
-	decryptor := keystorev4.New()
-	enc, err := decryptor.Decrypt(keystoreFile.Crypto, ks.password)
-	if err != nil {
-		return ErrKeystoreDecryption
-	}
-	store := &keyStoreAux{}
-	if err := json.Unmarshal(enc, store); err != nil {
-		return err
-	}
-	return ks.fromAux(store)
-}
-
-// saveStoreAndReInitialize saves the keystore to disk and re-initializes the keystore from file.
-func (ks *KeyStore) saveStoreAndReInitialize() error {
-	// Save the copy to disk
-	keystoreRepr, err := createKeystoreRepresentation(ks.toAux(), ks.password)
-	if err != nil {
-		return err
-	}
-	encodedData, err := json.MarshalIndent(keystoreRepr, "", "\t")
-	if err != nil {
-		return err
-	}
-	// A temporary hack to avoid useless keystore file overwrites until persistence logic is
-	// implemented over keystore, ref. https://github.com/bane-labs/go-ethereum/issues/388.
-	if ks.path != "" {
-		err = writeFileAtPath(filepath.Dir(ks.path), filepath.Base(ks.path), encodedData)
-		if err != nil {
-			return err
-		}
-
-		// ReInitialize and update the memory
-		err = ks.initializeKeystoreFromFile()
-	} else {
-		err = ks.initializeKeystoreFromBytes(encodedData)
-	}
-
-	return err
 }
 
 // createKeystoreRepresentation is a pure function that takes a keystore and password and returns the encrypted formatted json version for file writing.

--- a/antimev/keystore_test.go
+++ b/antimev/keystore_test.go
@@ -35,12 +35,11 @@ type account struct {
 // a copy if modification is needed since some tests rely on the order of accounts.
 // Ref. https://github.com/bane-labs/go-ethereum/tree/bane-main/privnet/seven.
 var accounts = []account{
-	// Node 0, only used as a watch-only node
-	// {
-	// 	common.HexToAddress("0x5e6D9680428e6fe62a09BBb6AC23Df5bFE069AE8"),
-	// 	"",
-	// 	"400fe4a0c2d74a4fb5cab917cc0e344f6ad33916b8e0e2e3815d2943134ddbe7",
-	// },
+	{
+		common.HexToAddress("0x5e6D9680428e6fe62a09BBb6AC23Df5bFE069AE8"),
+		"",
+		"400fe4a0c2d74a4fb5cab917cc0e344f6ad33916b8e0e2e3815d2943134ddbe7",
+	},
 	{
 		common.HexToAddress("0x74f4effb0b538baec703346b03b6d9292f53a4cd"),
 		"fBfgE23FfqSVZRCGzFZbFvqabF3Ewvcg",

--- a/antimev/keystore_test.go
+++ b/antimev/keystore_test.go
@@ -13,6 +13,7 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/crypto/ecies"
 	"github.com/ethereum/go-ethereum/crypto/tpke"
+	"github.com/stretchr/testify/require"
 )
 
 // 7-CN privnet related constants.
@@ -95,9 +96,7 @@ func TestShare(t *testing.T) {
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		kss[i] = ks
 	}
 	// Ignore resharing and execute sharing
@@ -107,14 +106,9 @@ func TestShare(t *testing.T) {
 	}
 	for i := 0; i < size; i++ {
 		// No reshare to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		msgs, pvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = msgs
 		contract.sharePVSSes[i] = pvss
 	}
@@ -131,21 +125,15 @@ func TestShare(t *testing.T) {
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	for i := 0; i < size; i++ {
 		// Try finish DKG without resharing
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 }
 
@@ -161,9 +149,7 @@ func TestReshare(t *testing.T) {
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		kss[i] = ks
 	}
 	// Ignore resharing and execute sharing
@@ -175,14 +161,9 @@ func TestReshare(t *testing.T) {
 	}
 	for i := 0; i < size; i++ {
 		// No reshare to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		msgs, pvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = msgs
 		contract.sharePVSSes[i] = pvss
 	}
@@ -199,36 +180,23 @@ func TestReshare(t *testing.T) {
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	// Finalize dkg
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 	// Execute resharing this time
 	for i := 0; i < size; i++ {
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		rMsgs, rPvss, err := kss[i].DKGReshare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		sMsgs, sPvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = sMsgs
 		contract.sharePVSSes[i] = sPvss
 		contract.reshareMsgs[i] = rMsgs
@@ -251,38 +219,30 @@ func TestReshare(t *testing.T) {
 	cmt = new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	// Check sharing and resharing
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 }
 
 func TestGroupChange(t *testing.T) {
 	dir := t.TempDir()
 	// Init keystores
-	pubs := make([]*ecies.PublicKey, len(accounts))
-	addrs := make([]common.Address, len(accounts))
-	kss := make([]*KeyStore, len(accounts))
-	for i := 0; i < len(accounts); i++ {
+	pubs := make([]*ecies.PublicKey, size+1)
+	addrs := make([]common.Address, size+1)
+	kss := make([]*KeyStore, size+1)
+	for i := 0; i < size+1; i++ {
 		key, _ := crypto.HexToECDSA(accounts[i].msgPrivKey)
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		addrs[i] = accounts[i].addr
 		kss[i] = ks
 	}
@@ -293,18 +253,13 @@ func TestGroupChange(t *testing.T) {
 		shareMsgs:     make([][][]byte, size),
 		sharePVSSes:   make([][]byte, size),
 	}
-	for i := 0; i < len(accounts); i++ {
+	for i := 0; i < len(addrs); i++ {
 		// No resharing to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		// Sharing members, i ranges from 0 to 6
 		if i < size {
 			msgs, pvss, err := kss[i].DKGShare(pubs[:size])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.shareMsgs[i] = msgs
 			contract.sharePVSSes[i] = pvss
 		}
@@ -314,55 +269,38 @@ func TestGroupChange(t *testing.T) {
 	for i := 0; i < size; i++ {
 		for j := 0; j < size; j++ {
 			err := kss[i].ReceiveSecretShare(i+1, j+1, contract.shareMsgs[j], contract.sharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Aggregate pvss manually
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	// Finalize dkg
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 	err := kss[7].OnEpochChange(nil, encodePointG1(cmt), false)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	// Execute resharing this time
 	for i := 0; i < len(addrs); i++ {
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		// Resharing members
 		if i < size {
 			rMsgs, rPvss, err := kss[i].DKGReshare(pubs[1:])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.reshareMsgs[i] = rMsgs
 			contract.resharePVSSes[i] = rPvss
 		}
 		if i > 0 {
 			sMsgs, sPvss, err := kss[i].DKGShare(pubs[1:])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.shareMsgs[i-1] = sMsgs
 			contract.sharePVSSes[i-1] = sPvss
 		}
@@ -372,56 +310,42 @@ func TestGroupChange(t *testing.T) {
 		for j := 0; j < size; j++ {
 			// Messages from node 0~6 to node 1~7
 			err := kss[i].ReceiveSecretReshare(i, j+1, contract.reshareMsgs[j], contract.resharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			// Messages from node 1~7 to node 1~7
 			err = kss[i].ReceiveSecretShare(i, j+1, contract.shareMsgs[j], contract.sharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Aggregate pvss manually
 	cmt = new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	// Check sharing and resharing
 	err = kss[0].OnEpochChange(nil, encodePointG1(cmt), false)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	for i := 1; i <= size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i-1], encodePointG1(cmt), i != 0)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 }
 
 func TestRecover(t *testing.T) {
 	dir := t.TempDir()
 	// Init keystores
-	pubs := make([]*ecies.PublicKey, len(accounts))
-	addrs := make([]common.Address, len(accounts))
-	kss := make([]*KeyStore, len(accounts))
-	for i := 0; i < len(accounts); i++ {
+	pubs := make([]*ecies.PublicKey, size+1)
+	addrs := make([]common.Address, size+1)
+	kss := make([]*KeyStore, size+1)
+	for i := 0; i < size+1; i++ {
 		key, _ := crypto.HexToECDSA(accounts[i].msgPrivKey)
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		addrs[i] = accounts[i].addr
 		kss[i] = ks
 	}
@@ -435,16 +359,11 @@ func TestRecover(t *testing.T) {
 	}
 	for i := 0; i < len(addrs); i++ {
 		// No resharing to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		// Sharing members, i ranges from 0 to 6
 		if i < size {
 			msgs, pvss, err := kss[i].DKGShare(pubs[:size])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.shareMsgs[i] = msgs
 			contract.sharePVSSes[i] = pvss
 		}
@@ -454,47 +373,32 @@ func TestRecover(t *testing.T) {
 	for i := 0; i < size; i++ {
 		for j := 0; j < size; j++ {
 			err := kss[i].ReceiveSecretShare(i+1, j+1, contract.shareMsgs[j], contract.sharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Aggregate pvss manually
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	// Finalize dkg
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 	err := kss[7].OnEpochChange(nil, encodePointG1(cmt), false)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	// Execute resharing this time
 	for i := 0; i < len(addrs); i++ {
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		// Resharing members
 		if i < size {
 			rMsgs, rPvss, err := kss[i].DKGReshare(pubs[1:])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.reshareMsgs[i] = rMsgs
 			contract.resharePVSSes[i] = rPvss
 		}
@@ -503,51 +407,37 @@ func TestRecover(t *testing.T) {
 	for i := 1; i < len(addrs); i++ {
 		for j := 0; j < size-1; j++ {
 			err := kss[i].ReceiveSecretReshare(i, j+1, contract.reshareMsgs[j], contract.resharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Execute recovering this time, dead index 7, recover to validator 8
 	rIdxs := []int{7}
 	rPubs := []*ecies.PublicKey{pubs[7]}
 	for i := 0; i < len(addrs); i++ {
-		err := kss[i].OnRecoverPeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnRecoverPeriodStart()
+		require.NoError(t, err)
 		if i < 7 {
 			msgs, err := kss[i].DKGRecover(rIdxs, rPubs)
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 			contract.recoverMsgs[i] = msgs
 		}
 	}
 	// Send recover messages, broadcast to all nodes
 	for i := 0; i < size-1; i++ {
 		err := kss[7].ReceiveRecoverShare(7, i+1, contract.recoverMsgs[i][0], contract.sharePVSSes[size-1])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 	// Recover the lost resharing messages
 	msgs, pvss, err := kss[7].TryRecoverReshare(pubs[1:])
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	for i := 1; i < len(addrs); i++ {
 		err := kss[i].ReceiveSecretReshare(i, 7, msgs, pvss)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 	// Only check resharing
 	for i := 0; i < len(addrs); i++ {
 		err := kss[i].aggregateReshare(i != 0)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 }
 

--- a/antimev/signature_test.go
+++ b/antimev/signature_test.go
@@ -38,9 +38,7 @@ func TestThresholdSignature(t *testing.T) {
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		kss[i] = ks
 	}
 	// Ignore resharing and execute sharing
@@ -50,14 +48,9 @@ func TestThresholdSignature(t *testing.T) {
 	}
 	for i := 0; i < size; i++ {
 		// No reshare to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		msgs, pvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = msgs
 		contract.sharePVSSes[i] = pvss
 	}
@@ -74,20 +67,14 @@ func TestThresholdSignature(t *testing.T) {
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 
 	// Test functionality
@@ -95,9 +82,7 @@ func TestThresholdSignature(t *testing.T) {
 	shares := make(map[int]*tpke.SignatureShare)
 	for i := 0; i < size; i++ {
 		share, err := kss[i].SignShare(msg)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		shares[i+1] = share
 	}
 

--- a/antimev/tpke_test.go
+++ b/antimev/tpke_test.go
@@ -103,6 +103,29 @@ func TestTPKE(t *testing.T) {
 	}
 }
 
+const regenerate = false
+
+// TestInitKeyStores generates antimev keystores to privnets
+func TestInitKeyStores(t *testing.T) {
+	if !regenerate {
+		return
+	}
+	sizes := []int{1, 4, 7}
+	thresholds := []int{1, 3, 5}
+	folders := []string{"single", "four", "seven"}
+
+	for i := range sizes {
+		for j := 0; j <= sizes[i]; j++ {
+			key, _ := crypto.HexToECDSA(accounts[j].msgPrivKey)
+			ks := NewKeyStore(filepath.Join("../privnet/"+folders[i]+"/node"+fmt.Sprint(j), "antimev-keystore"))
+			err := ks.Init(accounts[j].addr, ecies.ImportECDSA(key), sizes[i], thresholds[i], accounts[j].pwd)
+			require.NoError(t, err)
+			err = ks.Persist()
+			require.NoError(t, err)
+		}
+	}
+}
+
 // TestGenerateEncryptedTx generates an encrypted transaction using 7-nodes
 func TestGenerateEncryptedTx(t *testing.T) {
 	require.Equal(t, 7, size, "refactor test if different number of CNs is needed")

--- a/antimev/tpke_test.go
+++ b/antimev/tpke_test.go
@@ -32,9 +32,7 @@ func TestTPKE(t *testing.T) {
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		kss[i] = ks
 	}
 	// Ignore resharing and execute sharing
@@ -44,14 +42,9 @@ func TestTPKE(t *testing.T) {
 	}
 	for i := 0; i < size; i++ {
 		// No reshare to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		msgs, pvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = msgs
 		contract.sharePVSSes[i] = pvss
 	}
@@ -59,37 +52,27 @@ func TestTPKE(t *testing.T) {
 	for i := 0; i < size; i++ {
 		for j := 0; j < size; j++ {
 			err := kss[i].ReceiveSecretShare(i+1, j+1, contract.shareMsgs[j], contract.sharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Aggregate pvss manually
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 
 	// Encrypt
 	msg := []byte("some data that is more than 105 bytes in length: pizza pizza pizza pizza pizza pizza pizza pizza pizza pizza pizza pizza pizza")
 	encryptedKey, encryptedMsg, err := kss[0].Encrypt(msg)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 
 	// Verify ciphertext
 	if err := encryptedKey.Verify(); err != nil {
@@ -106,17 +89,13 @@ func TestTPKE(t *testing.T) {
 	shares := make(map[int][]*tpke.DecryptionShare)
 	for i := 0; i < size-2; i++ {
 		share, err := kss[i].DecryptWithShare([]*tpke.CipherText{encryptedKey})
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		shares[i+1] = share
 	}
 
 	// Decrypt
 	results, err := kss[0].AggregateAndDecryptWithShare([]*tpke.CipherText{encryptedKey}, [][]byte{encryptedMsg}, shares)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	for i := 0; i < len(msg); i++ {
 		if msg[i] != results[0][i] {
 			t.Fatalf("decryption failed.")
@@ -214,9 +193,7 @@ func TestBenchmark(t *testing.T) {
 		pubs[i] = &ecies.ImportECDSA(key).PublicKey
 		ks := NewKeyStore(filepath.Join(dir, "antimev-keystore"+fmt.Sprint(i)))
 		err := ks.Init(accounts[i].addr, ecies.ImportECDSA(key), size, threshold, accounts[i].pwd)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		kss[i] = ks
 	}
 	// Ignore resharing and execute sharing
@@ -226,14 +203,9 @@ func TestBenchmark(t *testing.T) {
 	}
 	for i := 0; i < size; i++ {
 		// No reshare to handle
-		err := kss[i].OnSharePeriodStart()
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		kss[i].OnSharePeriodStart()
 		msgs, pvss, err := kss[i].DKGShare(pubs)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		contract.shareMsgs[i] = msgs
 		contract.sharePVSSes[i] = pvss
 	}
@@ -241,29 +213,21 @@ func TestBenchmark(t *testing.T) {
 	for i := 0; i < size; i++ {
 		for j := 0; j < size; j++ {
 			err := kss[i].ReceiveSecretShare(i+1, j+1, contract.shareMsgs[j], contract.sharePVSSes[j])
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
+			require.NoError(t, err)
 		}
 	}
 	// Aggregate pvss manually
 	cmt := new(bls12381.G1Affine).ScalarMultiplicationBase(big.NewInt(0))
 	for i := 0; i < size; i++ {
 		p, err := new(tpke.PVSS).Decode(contract.sharePVSSes[i], size, threshold)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		pg1, err := decodePointG1(p.GetCommitment().Encode()[:128])
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		cmt = new(bls12381.G1Affine).Add(cmt, pg1)
 	}
 	for i := 0; i < size; i++ {
 		err := kss[i].OnEpochChange(contract.sharePVSSes[i], encodePointG1(cmt), true)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 	}
 
 	// Build a 1MB script
@@ -282,26 +246,20 @@ func TestBenchmark(t *testing.T) {
 		go parallelCTVerify(encryptedSeeds[i], ch)
 	}
 	_, _, err := messageHandler(ch, sampleAmount)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 
 	// Generate shares
 	shares := make(map[int][]*tpke.DecryptionShare)
 	for i := 0; i < size; i++ {
 		share, err := kss[i].DecryptWithShare(encryptedSeeds)
-		if err != nil {
-			t.Fatalf(err.Error())
-		}
+		require.NoError(t, err)
 		shares[i+1] = share
 	}
 
 	// Decrypt
 	t1 := time.Now()
 	results, err := kss[0].AggregateAndDecryptWithShare(encryptedSeeds, encryptedMsgs, shares)
-	if err != nil {
-		t.Fatalf(err.Error())
-	}
+	require.NoError(t, err)
 	t.Logf("threshold decryption time: %v", time.Since(t1))
 	for i := 0; i < sampleAmount; i++ {
 		if !bytes.Equal(results[i], script) {

--- a/cmd/geth/antimevcmd.go
+++ b/cmd/geth/antimevcmd.go
@@ -173,21 +173,19 @@ func antimevStatus(ctx *cli.Context) error {
 	ks := makeAntimevKeystore(ctx)
 	unlockKeyStore(ks, utils.MakeAntiMEVPasswordList(ctx))
 	var cpkStr, lpkStr string
-	cpk, err := ks.CurrentGlobalPubKey()
+	cpk, err := ks.GlobalPublicKey()
 	if err == nil {
 		cpkStr = hex.EncodeToString(cpk.Bytes())
 	}
-	lpk, err := ks.LastGlobalPubKey()
+	lpk, err := ks.LastGlobalPublicKey()
 	if err == nil {
 		lpkStr = hex.EncodeToString(lpk.Bytes())
 	}
 	fmt.Printf("Antimev keystore status:\n")
 	fmt.Printf("- Message public key: {%s}\n", ks.MessagePubKey())
-	fmt.Printf("- Round: {%d}\n", ks.Round())
-	fmt.Printf("- Last round reshared: {%t}\n", ks.HasReshared())
-	fmt.Printf("- This round shared: {%t}\n", ks.HasShared())
-	fmt.Printf("- This round resharing: {%t}\n", ks.IsResharing())
-	fmt.Printf("- Next round sharing: {%t}\n", ks.IsSharing())
+	fmt.Printf("- Successful rounds: {%d}\n", ks.Round())
+	fmt.Printf("- Resharing: {%t}\n", ks.IsResharing())
+	fmt.Printf("- Sharing: {%t}\n", ks.IsSharing())
 	fmt.Printf("- Current global key: {%s}\n", cpkStr)
 	fmt.Printf("- Last global key: {%s}\n", lpkStr)
 	return nil
@@ -221,6 +219,10 @@ func antimevInit(ctx *cli.Context) error {
 	if err != nil {
 		utils.Fatalf("Failed to init antimev keystore: %v", err)
 	}
+	err = ks.Persist()
+	if err != nil {
+		utils.Fatalf("Failed to persist antimev keystore: %v", err)
+	}
 	path, err := ks.Path()
 	if err != nil {
 		utils.Fatalf("Failed to get antimev keystore path: %v", err)
@@ -238,7 +240,8 @@ func antimevUpdate(ctx *cli.Context) error {
 	ks := makeAntimevKeystore(ctx)
 	unlockKeyStore(ks, utils.MakeAntiMEVPasswordList(ctx))
 	newPassword := utils.GetPassPhraseWithList("Please give a new password. Do not forget this password.", true, 0, nil)
-	if err := ks.Update(newPassword); err != nil {
+	ks.UpdatePassword(newPassword)
+	if err := ks.Persist(); err != nil {
 		utils.Fatalf("Could not update the keystore: %v", err)
 	}
 	return nil
@@ -258,7 +261,8 @@ func antimevReset(ctx *cli.Context) error {
 	if confirm {
 		unlockKeyStore(ks, utils.MakeAntiMEVPasswordList(ctx))
 		// Reset the keystore to the oldest possible state before any DKG
-		if err := ks.Reset(0); err != nil {
+		ks.Reset(0)
+		if err := ks.Persist(); err != nil {
 			utils.Fatalf("Could not reset the keystore: %v", err)
 		}
 	}

--- a/consensus/dbft/dbft.go
+++ b/consensus/dbft/dbft.go
@@ -2521,14 +2521,10 @@ func (c *DBFT) getGlobalPublicKey(h *types.Header, s *state.StateDB) (*tpke.Publ
 	// original keystore modification.
 	snapshot := c.dkgSnapshot.Copy()
 	c.lock.RLock()
-	ks := c.amevKeystore
+	keystore := c.amevKeystore.Copy()
 	c.lock.RUnlock()
-	keystore, err := ks.Copy()
-	if err != nil {
-		return nil, fmt.Errorf("failed to create keystore backup: %w", err)
-	}
 
-	err = c.handleDKG(snapshot, keystore, h, s.Copy(), true)
+	err := c.handleDKG(snapshot, keystore, h, s.Copy(), true)
 	if err != nil {
 		return nil, fmt.Errorf("failed to handle DKG at %d: %w", h.Number.Uint64(), err)
 	}

--- a/consensus/dbft/dkg.go
+++ b/consensus/dbft/dkg.go
@@ -143,7 +143,9 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		if err != nil {
 			return fmt.Errorf("failed to change keystore epoch, err: %v", err)
 		}
-		log.Info("DKG reached targetHeight", "round", snapshot.Round, "currentHeight", currentHeight, "aggregatedCommitment", hex.EncodeToString(aggregatedCommitment))
+		if !suspended {
+			log.Info("DKG reached targetHeight", "round", snapshot.Round, "currentHeight", currentHeight, "aggregatedCommitment", hex.EncodeToString(aggregatedCommitment))
+		}
 		snapshot.reset()
 	}
 
@@ -157,8 +159,10 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		if err != nil {
 			return fmt.Errorf("failed to new DKG snapshot, err: %v", err)
 		}
-		log.Info("DKG info", "roundNumber", snapshot.Round, "epochStartHeight", snapshot.EpochStartHeight, "epochDuration", epochDuration,
-			"sharePeriodDuration", sharePeriodDuration, "consensusList", snapshot.CurrentCNs)
+		if !suspended {
+			log.Info("DKG info", "roundNumber", snapshot.Round, "epochStartHeight", snapshot.EpochStartHeight, "epochDuration", epochDuration,
+				"sharePeriodDuration", sharePeriodDuration, "consensusList", snapshot.CurrentCNs)
+		}
 	}
 	// Compute periods based on realtime data, in case there is an update in governanace contract
 	targetHeight := snapshot.EpochStartHeight + epochDuration
@@ -221,7 +225,9 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		if err := keystore.OnEpochChange(selfPvss, aggregatedCommitment, indexOfSharing > 0); err != nil {
 			return fmt.Errorf("failed to sync shared DKG, err: %v", err)
 		}
-		log.Info("DKG sync to", "round", snapshot.Round-1, "currentHeight", currentHeight, "aggregatedCommitment", hex.EncodeToString(aggregatedCommitment))
+		if !suspended {
+			log.Info("DKG sync to", "round", snapshot.Round-1, "currentHeight", currentHeight, "aggregatedCommitment", hex.EncodeToString(aggregatedCommitment))
+		}
 	}
 
 	// DKG checkpoint handling, also syncs the dkg process if keystore is out-of-date

--- a/consensus/dbft/dkg.go
+++ b/consensus/dbft/dkg.go
@@ -184,24 +184,18 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 	if keystoreRound == int(snapshot.Round)-1 && currentHeight < shareStartHeight && keystore.IsSharing() {
 		// Here only revert before the new share period, if the node wake up when share,
 		// then data will be overwritten
-		if err := keystore.RevertRound(); err != nil {
-			return err
-		}
+		keystore.RevertRound()
 	}
 	// Sync if keystore needs a sync and contract has at least 1 round successful DKG
 	if keystoreRound < int(snapshot.Round)-1 {
 		// If keystore is late for more than 1 round, than reset it to the last one,
 		// past data are not helpful anymore
 		if keystoreRound < int(snapshot.Round)-2 {
-			if err := keystore.Reset(int(snapshot.Round) - 2); err != nil {
-				return err
-			}
+			keystore.Reset(int(snapshot.Round) - 2)
 		}
 		// If keystore is in a sharing state already, then regard it has a valid secret
 		if !keystore.IsSharing() {
-			if err := keystore.OnSharePeriodStart(); err != nil {
-				return fmt.Errorf("failed to sync shared DKG, err: %v", err)
-			}
+			keystore.OnSharePeriodStart()
 		}
 		// If is a member of current consensus, then try sync secrets
 		indexOfSharing := slices.Index(snapshot.CurrentCNs, amevAddress) + 1
@@ -239,29 +233,26 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		}
 		// Prepare and start a DKG
 		if !keystore.IsSharing() {
-			if err = keystore.OnSharePeriodStart(); err != nil {
-				return fmt.Errorf("failed to start new DKG, err: %v", err)
-			}
+			keystore.OnSharePeriodStart()
 		}
 		// If the period finished, then skip sending a transaction
-		if currentHeight < recoverStartHeight {
+		if !suspended && currentHeight < recoverStartHeight {
+			// If is a member of current consensus, try reshare but give up if error
+			indexOfResharing := slices.Index(snapshot.CurrentCNs, amevAddress) + 1
+			receiverMessageKeys, err := getMessagePubkeys(c.backend, snapshot.PendingCNs, state, h)
+			if indexOfResharing > 0 && snapshot.Round > 1 {
+				if err = c.taskReshare(keystore, receiverMessageKeys, currentHeight, recoverStartHeight, watchList); err != nil {
+					return fmt.Errorf("failed to task DKG reshare, err: %v", err)
+				}
+			}
 			// If is a member of pending consensus
 			indexOfSharing := slices.Index(snapshot.PendingCNs, amevAddress) + 1
-			receiverMessageKeys, err := getMessagePubkeys(c.backend, snapshot.PendingCNs, state, h)
 			if err != nil {
 				return fmt.Errorf("failed to get message keys, err: %v", err)
 			}
-			if !suspended && indexOfSharing > 0 {
+			if indexOfSharing > 0 {
 				if err = c.taskShare(keystore, receiverMessageKeys, currentHeight, recoverStartHeight, watchList); err != nil {
 					return fmt.Errorf("failed to task DKG share, err: %v", err)
-				}
-			}
-			// If is a member of current consensus, try reshare but give up if error
-			indexOfResharing := slices.Index(snapshot.CurrentCNs, amevAddress) + 1
-			if !suspended && indexOfResharing > 0 && snapshot.Round > 1 {
-				if err = c.taskReshare(keystore, receiverMessageKeys, currentHeight, recoverStartHeight, watchList); err != nil {
-					snapshot.ShareTasked = true
-					return fmt.Errorf("failed to task DKG reshare, err: %v", err)
 				}
 			}
 		}
@@ -274,49 +265,41 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		if err != nil {
 			return fmt.Errorf("failed to check if sharing is ready: %v", err)
 		}
-		if !ready {
-			snapshot.RecoverTasked = true
-			log.Warn("DKG sharing is not ready, skip recover")
-			return nil
-		}
-		// If share is ready, pending consensus nodes should ReceiveSecretShare
-		indexOfSharing := slices.Index(snapshot.PendingCNs, amevAddress) + 1
-		if indexOfSharing > 0 {
-			if err := c.syncThisRoundSecrets(snapshot, keystore, indexOfSharing, state, h); err != nil {
-				return fmt.Errorf("failed to sync sharing DKG, err: %v", err)
-			}
-		}
-		snapshot.IndexNeedRecover, err = getIndexCurrentNeedRecovering(c.backend, state, h)
-		if err != nil {
-			return fmt.Errorf("failed to fecth index need recovering, err: %v", err)
-		}
-		// If the period finished, then skip sending a transaction
-		if len(snapshot.IndexNeedRecover) > 0 {
-			// Only indexesNeedRecover <= (consensusSize - threshold) can recover
-			threshold := consensusSize - (consensusSize-1)/3
-			if len(snapshot.IndexNeedRecover) > int(consensusSize-threshold) {
-				snapshot.RecoverTasked = true
-				log.Warn("DKG resharing doesn't meet recoverable threshold, skip recover")
-				return nil
-			}
-			if !keystore.IsRecovering() {
-				if err := keystore.OnRecoverPeriodStart(); err != nil {
-					return fmt.Errorf("failed to start DKG recover, err: %v", err)
+		if ready {
+			// If share is ready, pending consensus nodes should ReceiveSecretShare
+			indexOfSharing := slices.Index(snapshot.PendingCNs, amevAddress) + 1
+			if indexOfSharing > 0 {
+				if err := c.syncThisRoundSecrets(snapshot, keystore, indexOfSharing, state, h); err != nil {
+					return fmt.Errorf("failed to sync sharing DKG, err: %v", err)
 				}
 			}
-			if !suspended && currentHeight < recoverCheckHeight {
-				// Send recover tx from current consensus node
-				indexOfResharing := slices.Index(snapshot.CurrentCNs, amevAddress) + 1
-				if indexOfResharing > 0 {
-					pubs := make([]*ecies.PublicKey, len(snapshot.IndexNeedRecover))
-					for i, index := range snapshot.IndexNeedRecover {
-						pubs[i], err = getMessagePubkey(c.backend, snapshot.PendingCNs[index-1], state, h)
-						if err != nil {
-							return fmt.Errorf("failed to get message keys for recovering, err: %v", err)
-						}
+			snapshot.IndexNeedRecover, err = getIndexCurrentNeedRecovering(c.backend, state, h)
+			if err != nil {
+				return fmt.Errorf("failed to fetch index need recovering, err: %v", err)
+			}
+			// If the period finished, then skip sending a transaction
+			if len(snapshot.IndexNeedRecover) > 0 {
+				// Only indexesNeedRecover <= (consensusSize - threshold) can recover
+				threshold := consensusSize - (consensusSize-1)/3
+				if len(snapshot.IndexNeedRecover) <= int(consensusSize-threshold) {
+					if !keystore.IsRecovering() {
+						keystore.OnRecoverPeriodStart()
 					}
-					if err := c.taskRecover(keystore, snapshot.IndexNeedRecover, pubs, currentHeight, recoverCheckHeight, watchList); err != nil {
-						return fmt.Errorf("failed to task DKG recover, err: %v", err)
+					if !suspended && currentHeight < recoverCheckHeight {
+						// Send recover tx from current consensus node
+						indexOfResharing := slices.Index(snapshot.CurrentCNs, amevAddress) + 1
+						if indexOfResharing > 0 {
+							pubs := make([]*ecies.PublicKey, len(snapshot.IndexNeedRecover))
+							for i, index := range snapshot.IndexNeedRecover {
+								pubs[i], err = getMessagePubkey(c.backend, snapshot.PendingCNs[index-1], state, h)
+								if err != nil {
+									return fmt.Errorf("failed to get message keys for recovering, err: %v", err)
+								}
+							}
+							if err := c.taskRecover(keystore, snapshot.IndexNeedRecover, pubs, currentHeight, recoverCheckHeight, watchList); err != nil {
+								return fmt.Errorf("failed to task DKG recover, err: %v", err)
+							}
+						}
 					}
 				}
 			}
@@ -329,21 +312,23 @@ func (c *DBFT) handleDKG(snapshot *Snapshot, keystore *antimev.KeyStore, h *type
 		// Send reshareRecovered at height recoverStartHeigh+c.shareDuration/2
 		// Only index in indexsNeedRecover and pending consensus node need to call reshareRecovered
 		indexOfSharing := slices.Index(snapshot.PendingCNs, amevAddress) + 1
-		if indexOfSharing > 0 && currentHeight < targetHeight && slices.Contains(snapshot.IndexNeedRecover, uint64(indexOfSharing)) {
+		if !suspended && indexOfSharing > 0 && currentHeight < targetHeight && slices.Contains(snapshot.IndexNeedRecover, uint64(indexOfSharing)) {
 			if err := c.syncRecoveredSecrets(snapshot, keystore, indexOfSharing, state, h); err != nil {
 				return fmt.Errorf("failed to sync recovering DKG, err: %v", err)
 			}
-			if !suspended {
-				receiverMessageKeys, err := getMessagePubkeys(c.backend, snapshot.PendingCNs, state, h)
-				if err != nil {
-					return fmt.Errorf("failed to fecth message keys, err: %v", err)
-				}
-				if err := c.taskReshareRecover(keystore, receiverMessageKeys, currentHeight, targetHeight, watchList); err != nil {
-					return fmt.Errorf("failed to task DKG reshare recover, err: %v", err)
-				}
+			receiverMessageKeys, err := getMessagePubkeys(c.backend, snapshot.PendingCNs, state, h)
+			if err != nil {
+				return fmt.Errorf("failed to fecth message keys, err: %v", err)
+			}
+			if err := c.taskReshareRecover(keystore, receiverMessageKeys, currentHeight, targetHeight, watchList); err != nil {
+				return fmt.Errorf("failed to task DKG reshare recover, err: %v", err)
 			}
 		}
 		snapshot.ReshareRecoverTasked = true
+	}
+
+	if err := keystore.Persist(); err != nil {
+		return fmt.Errorf("failed to persist keystore, err: %v", err)
 	}
 	return nil
 }

--- a/crypto/tpke/poly.go
+++ b/crypto/tpke/poly.go
@@ -35,6 +35,17 @@ func (p *Poly) evaluate(x *big.Int) *big.Int {
 	return result
 }
 
+// Copy returns a deep copy of Poly
+func (p *Poly) Copy() *Poly {
+	coeff := make([]*big.Int, len(p.coeff))
+	for i := range coeff {
+		coeff[i] = new(big.Int).Set(p.coeff[i])
+	}
+	return &Poly{
+		coeff: coeff,
+	}
+}
+
 func (p *Poly) AddAssign(op *Poly) {
 	pLen := len(p.coeff)
 	opLen := len(op.coeff)

--- a/crypto/tpke/private_key.go
+++ b/crypto/tpke/private_key.go
@@ -30,6 +30,13 @@ func NewPrivateKey(secretShares []*big.Int) *PrivateKey {
 	}
 }
 
+// Copy returns a deep copy of PrivateKey
+func (sk *PrivateKey) Copy() *PrivateKey {
+	return &PrivateKey{
+		fr: new(big.Int).Set(sk.fr),
+	}
+}
+
 func (sk *PrivateKey) Bytes() []byte {
 	fe := new(fr.Element).SetBigInt(sk.fr)
 	b := fe.Bytes()

--- a/crypto/tpke/public_key.go
+++ b/crypto/tpke/public_key.go
@@ -64,6 +64,13 @@ func NewPublicKeyFromBytes(b []byte) (*PublicKey, error) {
 	return pk, nil
 }
 
+// Copy returns a deep copy of PublicKey
+func (pk *PublicKey) Copy() *PublicKey {
+	return &PublicKey{
+		pg1: new(bls12381.G1Affine).Set(pk.pg1),
+	}
+}
+
 // Bytes serializes PublicKey into byte slice using compressed [bls12381.G1Affine]
 // representation. The resulting byte slice has the length of
 // [PublicKeyLen].

--- a/crypto/tpke/secrect.go
+++ b/crypto/tpke/secrect.go
@@ -24,6 +24,13 @@ func RecoverSecret(is []int, fis []*big.Int) *Secret {
 	}
 }
 
+// Copy returns a deep copy of Secret
+func (s *Secret) Copy() *Secret {
+	return &Secret{
+		poly: s.poly.Copy(),
+	}
+}
+
 func (s *Secret) ToBigIntArray() []*big.Int {
 	return s.poly.coeff
 }

